src/render/mod.rs

```diff
@@ -193,19 +202,71 @@ gfx_defines! {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub(crate) struct InstanceCacheKey {
+    pub(crate) material_id: usize,
+    pub(crate) geometry_id: h::Buffer<back::Resources, Vertex>,
+}
+
+impl Instance {
+    #[inline]
+    fn basic(
+        mx_world: [[f32; 4]; 4],
+        color: u32,
+        uv_range: [f32; 4],
+        param: f32,
+    ) -> Instance {
+        Instance {
+            world0: mx_world[0],
+            world1: mx_world[1],
+            world2: mx_world[2],
+            world3: mx_world[3],
+            color: {
+                let rgb = color::to_linear_rgb(color);
+                [rgb[0], rgb[1], rgb[2], 0.0]
+            },
+            mat_params: [param, 0.0, 0.0, 0.0],
+            uv_range,
+        }
+    }
+
+    #[inline]
+    fn pbr(mx_world: [[f32; 4]; 4]) -> Instance {
+        Instance {
+            world0: mx_world[0],
+            world1: mx_world[1],
+            world2: mx_world[2],
+            world3: mx_world[3],
+            color: [0.0; 4],
+            mat_params: [0.0; 4],
+            uv_range: [0.0; 4],
+        }
+    }
+}
+
 //TODO: private fields?
 #[derive(Clone, Debug)]
 pub(crate) struct GpuData {
     pub slice: gfx::Slice<back::Resources>,
-    pub vertices: gfx::handle::Buffer<back::Resources, Vertex>,
-    pub constants: gfx::handle::Buffer<back::Resources, Locals>,
+    pub vertices: h::Buffer<back::Resources, Vertex>,
+    pub instances: h::Buffer<back::Resources, Instance>,
     pub pending: Option<DynamicData>,
+    pub use_instancing: bool,
+    pub instance_cache_key: InstanceCacheKey,
```

Comment:
could we have `Option<InstanceCacheKey>` for clearer semantics than a separate boolean flag?